File: nim/examples/basic_example.nim
## Basic example demonstrating the GitHub Copilot SDK for Nim.
##
## Shows how to create a client, start a session, register a tool,
## send a message, and read the response.

import std/[json, asyncdispatch, os]

# Adjust import path for running from nim/examples/
import ../src/copilot_sdk

proc main() {.async.} =
  # Create and start the client
  let client = newCopilotClient(newClientConfig(
    cliPath = getEnv("COPILOT_CLI_PATH", ""),
  ))

  await client.start()
  echo "Client connected."

  # Register a simple tool
  await client.defineTool(
    "get_weather",
    "Returns the current weather for a city",
    @[
      newToolParameter("city", "string", "City name"),
    ],
    proc(params: JsonNode): string =
      let city = params{"city"}.getStr("unknown")
      $(%*{"city": city, "temperature": "22C", "condition": "sunny"}),
  )

  # Create a session with a system prompt
  let session = await client.createSession(newSessionConfig(
    systemPrompt = "You are a helpful assistant with access to weather tools.",
  ))
  echo "Session created: ", session.sessionId

  # Subscribe to streaming events
  session.on("assistant.message_delta", proc(event: SessionEvent) =
    if event.data.hasKey("delta"):
      stdout.write(event.data["delta"].getStr())
  )

  # Send a message and wait for the complete response
  let response = await session.sendAndWait(newMessageOptions(
    "What is the weather like in Tokyo?",
  ))
  echo "\n\nFull response: ", response.message

  # Clean up
  await session.disconnect()
  client.stop()
  echo "Done."

waitFor main()

File: fortran/examples/basic_example.f90
! -------------------------------------------------------------------------------------
!  Copyright (c) Microsoft Corporation. All rights reserved.
! -------------------------------------------------------------------------------------

!> Basic example: create a client, open a session, and send a message.
program basic_example
  use copilot_client_module
  use copilot_types
  use copilot_session_module
  implicit none

  type(copilot_client)        :: client
  type(copilot_client_options) :: opts
  type(session_config)        :: config
  type(copilot_session_type)  :: session
  type(send_options)          :: sopts
  type(send_and_wait_result)  :: result
  integer                     :: ios

  ! ------------------------------------------------------------------
  ! 1. Configure and create the client
  ! ------------------------------------------------------------------
  opts%log_level = 'info'
  ! opts%cli_path = '/usr/local/bin/copilot'   ! uncomment to override
  call copilot_client_create(client, opts)

  ! ------------------------------------------------------------------
  ! 2. Create a session
  ! ------------------------------------------------------------------
  config%model = 'gpt-4'
  config%streaming = .true.
  call copilot_create_session(client, config, session, ios)
  if (ios /= 0) then
    print *, 'ERROR: failed to create session'
    call copilot_stop(client)
    stop 1
  end if

  print *, 'Session created: ', session%session_id

  ! ------------------------------------------------------------------
  ! 3. Send a message and wait for the response
  ! ------------------------------------------------------------------
  sopts%message = 'What is 2 + 2?'
  call session%send_and_wait(sopts, result)

  if (result%is_error) then
    print *, 'ERROR: ', result%error_message
  else
    print *, 'Response: ', result%message
  end if

  ! ------------------------------------------------------------------
  ! 4. Clean up
  ! ------------------------------------------------------------------
  call session%destroy(ios)
  call copilot_stop(client)

  print *, 'Done.'
end program basic_example
